### PR TITLE
Fix WatchedResource data race

### LINKED ISSUE #51685: data race
https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/51684/unit-tests_istio/1805245436798177280

### PR DESCRIPTION
**Pull request by wulianglongrd**
fix #51685

### CODE REVIEW

File: pkg/xds/server.go [hunk 1]
```diff
@@ -91,10 +92,21 @@ type WatchedResource struct {
...
 	// LastResources tracks the contents of the last push.
-	// This field is extremely expensive to maintain and is typically disabled
+	// This field is extremely expensive to maintain and is typically disabled.
+	// It is only used for testing now.
 	LastResources Resources
 }
 
+func (r *WatchedResource) DeepCopy() *WatchedResource {
+	out := *r
+	if len(r.ResourceNames) > 0 {
```
Comment (howardjohn) [inline pkg/xds/server.go:102]: I am worried about this cost. Especially since its to support a debug interface, which doesn't even depend on resourceNames

**Review by hzxuzhonghu — CHANGES_REQUESTED** [review]
I am thinking we should not fix this kind of race. It doenot influence istio work correctly. But introducing performance issue.

**Comment by wulianglongrd** [comment]
I'm not sure if this fix makes sense. In theory there are data races in the core process (but I haven't really encountered them in production), for example:

read：

https://github.com/istio/istio/blob/27a44759d96d7e532a62321c8a2e2f33f6920e1f/pilot/pkg/xds/ads.go#L635


write：

https://github.com/istio/istio/blob/27a44759d96d7e532a62321c8a2e2f33f6920e1f/pkg/xds/server.go#L405

**Comment by hzxuzhonghu** [comment]
I think they cannot run concurrently, donot they run in a goroutine.

**Comment by wulianglongrd** [comment]
> I think they cannot run concurrently, donot they run in a goroutine.

You are correct, my example above is incorrect.

**Comment by howardjohn** [comment]
Its only the debug interfaces that do the concurrent access

**Pull request closed without merging**